internal/e2e/e2eapi/e2eapi_test.go

```diff
@@ -116,4 +120,56 @@ func TestDo(t *testing.T) {
 		}
 		require.ErrorContains(t, err, "unsupported protocol")
 	}
+
+	func() {
+		hr := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			w.WriteHeader(http.StatusNoContent)
+		})
+
+		ts := httptest.NewServer(hr)
+		defer ts.Close()
+
+		err := Do(ctx, http.MethodPost, ts.URL, nil, nil)
+		if err != nil {
+			t.Fatalf("exp nil err; got %v", err)
+		}
+	}()
+
+	for _, statusCode := range []int{http.StatusBadRequest, http.StatusOK} {
+		func() {
+			sentinel := errors.New("sentinel")
+			rtFn := roundTripperFunc(func(req *http.Request) (*http.Response, error) {
+				res, err := http.DefaultClient.Do(req)
+				if err != nil {
+					return nil, err
+				}
+				res.Body = io.NopCloser(iotest.ErrReader(sentinel))
+				return res, nil
+			})
+
+			prev := defaultClient
+			defer func() {
+				defaultClient = prev
+			}()
+			defaultClient = new(http.Client)
+			defaultClient.Transport = rtFn
+
+			hr := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				w.WriteHeader(statusCode)
+			})
+
+			ts := httptest.NewServer(hr)
+			defer ts.Close()
+
+			err := Do(ctx, http.MethodPost, ts.URL, nil, nil)
+			require.Error(t, err)
+			require.Equal(t, sentinel, err)
+		}()
+	}
```

Comment:
I'm reading this for 5 minutes and I can't see what is it doing? Can you clean it up so it's either more obvious, or add some comments.

Reply:
I've added detailed comments for this.